File: api/app/utils/saisi_pour.py
"""**« Saisi pour »** — au nom de qui une entrée est ouverte, et par qui.

## La notion

Le conseil syndical enregistre ce qu'un résident a signalé par téléphone, ou ce
qu'un intervenant extérieur a rapporté. L'**auteur** est alors celui qui a tapé ;
le **propriétaire** est celui que ça concerne.

Trois champs, et ils vont **toujours ensemble** :

| champ | ce qu'il porte |
|---|---|
| `saisi_pour_user_id` | un résident **inscrit** — nom et adresse à jour |
| `saisi_pour_nom` | une personne **extérieure** — ce que le rédacteur a saisi |
| `saisi_pour_email` | son adresse, **facultative** |

## 🔴 Pourquoi ce module (15/09/2026)

La notion vivait dans le seul circuit des tickets, écrite en quatre endroits —
le modèle, la création, la correction, la lecture. La demande de l'étendre aux
**actualités** et aux **événements** en aurait fait douze.

> « As-tu réalisé l'évolution d'ajout de la section “SAISI POUR” aux actualités
>   et Événements ? »

⚠️ Et la logique n'a rien d'évident : elle porte deux distinctions que trois
copies auraient perdues à des moments différents.

1. **Les trois champs voyagent ensemble, y compris à `null`.** C'est leur
   *présence* dans la charge utile qui dit d'écrire (`model_fields_set`), et
   c'est ce qui permet de revenir à « En mon nom » : sans elles, un `None`
   serait indistinguable d'un champ non transmis, et le choix resterait sans
   effet — en silence.
2. **La PRÉSENCE décide d'écrire, la COMPARAISON décide d'annoncer**
   (18/08/2026). Les deux étaient confondues : le formulaire envoyant toujours
   les trois champs, « Saisi pour modifié » apparaissait à chaque
   enregistrement, même quand personne n'y avait touché.

## Ce que ce module ne porte pas

**Qui est le propriétaire** — `utils/copie_auteur.proprietaire()`, déjà
générique : elle lit `saisi_pour_*` par `getattr` et retombe sur l'auteur pour
les objets qui n'en ont pas. Elle n'avait pas besoin de changer, et c'est le
signe qu'elle était juste.

**Qui a le droit** — `auth/deps`. Le filtre `est_cs` ci-dessous n'est pas une
autorisation : c'est la traduction d'une règle métier déjà tranchée par la
dépendance qui protège la route. Un résident qui enverrait ces champs les voit
simplement ignorés.
"""
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel

from sqlmodel import Field, Session, SQLModel


#: Les trois champs, nommés une fois — les parcourir plutôt que les énumérer.
CHAMPS = ("saisi_pour_user_id", "saisi_pour_nom", "saisi_pour_email")


class SaisiPourMixin(SQLModel):
    """Les trois colonnes, héritées plutôt que recopiées.

    ⚠️ **`saisi_pour_user_id` ne déclare PAS `foreign_key` ici**, et c'est une
    contrainte de SQLite, pas un oubli : une migration ne peut pas ajouter une
    colonne *avec* sa contrainte (`ALTER TABLE` la refuse, la migration crashe
    après avoir posé la colonne — c'est arrivé deux fois, 0117 et 0165). Si le
    modèle la déclarait quand même, une base **neuve** (`create_all`) et une base
    **migrée** porteraient deux schémas différents, et rien ne le dirait.
    `api/tests/test_migrations.py` refuse les deux.

    `Ticket` **redéclare** ce champ avec sa clé étrangère : sa colonne existe en
    base depuis l'origine, avec la contrainte. Le mixin décrit donc le socle, et
    l'entité qui en sait plus l'enrichit — sans que les trois autres aient à
    connaître ce détail.
    """

    saisi_pour_user_id: Optional[int] = Field(default=None)
    saisi_pour_nom: Optional[str] = None
    saisi_pour_email: Optional[str] = None


def valeurs(body, *, autorise: bool) -> dict:
    """Ce qu'il faut poser à la CRÉATION — les trois champs, ou trois `None`.

    :param autorise: le rédacteur a-t-il le droit de saisir pour autrui (le
        conseil syndical, l'administration) ? Sinon les trois valeurs sont
        ignorées — ce n'est pas le contrôle d'accès, qui vit dans `auth/deps`,
        mais sa traduction sur ces trois champs.
    """
    return {c: (getattr(body, c, None) if autorise else None) for c in CHAMPS}


def corriger(objet, body, envoye) -> list[str]:
    """Applique une CORRECTION, et dit s'il y a lieu de l'annoncer.

    :param envoye: `lambda body, champ: …` — le prédicat de l'appelant qui dit
        si un champ figure dans la charge utile reçue. Il reste chez lui : la
        façon de lire `model_fields_set` dépend du schéma, pas de cette notion.
    :returns: la liste des évolutions à tracer — vide si rien n'a changé.

    🔴 Les deux règles de l'en-tête sont ICI, et nulle part ailleurs : la
    présence décide d'écrire, la comparaison décide d'annoncer.
    """
    if not any(envoye(body, c) for c in CHAMPS):
        return []
    avant = tuple(getattr(objet, c) for c in CHAMPS)
    apres = tuple(getattr(body, c, None) for c in CHAMPS)
    for champ, valeur in zip(CHAMPS, apres):
        setattr(objet, champ, valeur)
    return ["Saisi pour modifié"] if avant != apres else []


def noms_derives(session: Session, objet) -> tuple[Optional[str], Optional[str]]:
    """Les **deux** noms dérivés d'un objet, en un seul calcul.

    :returns: `(proprietaire_nom, saisi_pour_affichage)`.

    ## 🔴 Pourquoi les rendre ensemble (21/09/2026, #1104)

    Ils viennent de la MÊME question — « à qui cet objet appartient-il ? » — et
    ne diffèrent que par le cas où personne n'est nommé : le propriétaire
    retombe sur l'auteur, l'affichage reste vide.

    Ils étaient pourtant composés à trois endroits et de deux façons : le ticket
    appelait `proprietaire()` puis en dérivait l'affichage, tandis que
    l'actualité et l'événement appelaient `affichage()`, qui refait le même
    appel. Résultat — **seul le ticket exposait `proprietaire_nom`**, et les
    cartes d'actualité et d'événement n'avaient aucun moyen d'afficher autre
    chose que le rédacteur. C'est la moitié serveur de #1104.

    ⚠️ La nuance entre les deux se lit ici, une fois. La répartir entre trois
    routeurs, c'est trois occasions de l'oublier — et elle l'a été deux fois.
    """
    #  🔴 Import DIFFÉRÉ, et c'est un cycle réel : `copie_auteur` a besoin des
    #  modèles, et les modèles ont besoin du mixin de ce fichier. Le poser en
    #  tête ferait échouer le DÉMARRAGE de l'API, pas un test.
    from app.utils.copie_auteur import proprietaire

    nom, _ = proprietaire(session, objet)
    nomme = bool(
        getattr(objet, "saisi_pour_user_id", None) or getattr(objet, "saisi_pour_nom", None)
    )
    return nom, (nom if nomme else None)


class SaisiPourEntree(BaseModel):
    """Ce qu'un formulaire ENVOIE — hérité par les schémas de création et de
    correction des trois entités.

    ⚠️ **Les trois champs sont `Optional` et sans valeur par défaut autre que
    `None`**, y compris à la création : c'est leur PRÉSENCE dans la charge utile
    qui dit d'écrire (`model_fields_set`), et c'est ce qui permet de revenir à
    « En mon nom ». Un défaut « intelligent » ici rendrait ce retour impossible.
    """

    saisi_pour_user_id: Optional[int] = None
    saisi_pour_nom: Optional[str] = None
    saisi_pour_email: Optional[str] = None


class SaisiPourSortie(BaseModel):
    """Ce qu'un écran REÇOIT — les trois champs, plus le nom déjà composé.

    ⚠️ `saisi_pour_affichage` n'est pas une redite des deux autres : il vaut
    `None` quand personne n'est nommé, là où `proprietaire()` retombe sur
    l'auteur. C'est lui qui décide si l'écran affiche « Saisi pour X » — le
    composer côté client demanderait de relire la règle des trois cas dans
    chaque écran.
    """

    saisi_pour_user_id: Optional[int] = None
    saisi_pour_nom: Optional[str] = None
    saisi_pour_email: Optional[str] = None
    saisi_pour_affichage: Optional[str] = None
    #  🔴 À QUI l'objet appartient — le « Saisi pour » s'il existe, l'auteur
    #  sinon. Déclaré ICI depuis le 21/09/2026 (#1104) : il ne vivait que dans
    #  `TicketRead`, si bien que l'actualité et l'événement ne pouvaient pas
    #  afficher autre chose que leur rédacteur. Les trois héritent déjà de cette
    #  classe — c'est ce qui rend la règle vraie partout d'une seule ligne.
    proprietaire_nom: Optional[str] = None

File: front/src/lib/saisi-pour.ts
/**
 * **« Saisi pour »** — au nom de qui une entrée est ouverte, côté écran.
 *
 * ## Pourquoi ce module (15/09/2026)
 *
 * La notion vivait dans le seul circuit des tickets. Son extension aux
 * **actualités** et aux **événements**, demandée à l'écran, aurait recopié deux
 * fois chacune de ses trois règles :
 *
 * 1. **quel mode** un objet déjà enregistré rouvre (`modeDepuis`) ;
 * 2. **ce qui part** vers l'API selon le mode (`lotSaisiPour`) ;
 * 3. quelles valeurs le formulaire remet à zéro.
 *
 * Aucune n'est évidente, et la deuxième est franchement contre-intuitive.
 *
 * ## ⚠️ Les trois champs partent TOUJOURS, y compris à `null`
 *
 * C'est leur **présence** dans la charge utile qui dit au serveur d'écrire
 * (`model_fields_set` côté Python, cf. `api/app/utils/saisi_pour.py`), et c'est
 * ce qui permet de revenir à « En mon nom » : sans elles, un `null` serait
 * indistinguable d'un champ non transmis, et le choix resterait sans effet — en
 * silence.
 *
 * ⚠️ Et le mode décide **lesquels** valent quelque chose : revenir d'« une
 * personne extérieure » à « un résident inscrit » doit effacer le nom saisi,
 * sinon un ancien destinataire survivrait à la personne qui le remplace.
 *
 * ## Le pendant serveur
 *
 * `api/app/utils/saisi_pour.py` — même notion, mêmes trois champs, et le mixin
 * dont héritent `Ticket`, `Publication` et `Evenement`. Les contextes de build
 * étant `./api` et `./front`, le partage d'un fichier est impossible (mémoire
 * `project_partage_front_api_impossible`) : ce sont deux écritures assumées
 * d'une même règle, et c'est le serveur qui tranche.
 */

/** En son nom, pour un résident inscrit, ou pour une personne extérieure. */
export type ModeSaisiPour = 'moi' | 'resident' | 'exterieur';

/** Ce qu'un objet déjà enregistré porte de « Saisi pour ». */
export interface PorteSaisiPour {
	saisi_pour_user_id?: number | null;
	saisi_pour_nom?: string | null;
	saisi_pour_email?: string | null;
}

/**
 * Le mode qu'un objet ROUVRE à l'édition.
 *
 * 🔴 L'ordre des deux tests n'est pas indifférent : un résident inscrit peut
 * avoir laissé un nom d'une saisie antérieure, et c'est l'identifiant qui fait
 * foi. L'inverse rouvrirait « personne extérieure » sur un ticket rattaché à un
 * compte.
 *
 * ⚠️ Rouvrir sur « En mon nom » par défaut effacerait le « Saisi pour » au
 * premier enregistrement — c'est pour cela que cette fonction existe plutôt
 * qu'un `'moi'` initial.
 */
export function modeDepuis(objet: PorteSaisiPour | null | undefined): ModeSaisiPour {
	if (objet?.saisi_pour_user_id) return 'resident';
	if (objet?.saisi_pour_nom) return 'exterieur';
	return 'moi';
}

/**
 * Ce qui part vers l'API — les **trois** champs, toujours.
 *
 * ⚠️ Ne jamais filtrer ce que cette fonction rend, ni l'étaler
 * conditionnellement : c'est la présence des trois clés qui permet d'effacer.
 */
export function lotSaisiPour(
	mode: ModeSaisiPour,
	userId: number | null,
	nom: string,
	email: string,
): Required<PorteSaisiPour> {
	return {
		saisi_pour_user_id: mode === 'resident' ? userId : null,
		saisi_pour_nom: mode === 'exterieur' ? nom.trim() || null : null,
		saisi_pour_email: mode === 'exterieur' ? email.trim() || null : null,
	};
}

/**
 * Les trois champs d'un formulaire, repris d'un objet existant ou vides.
 *
 * ⚠️ **Les trois sont toujours présents**, y compris pour une création : c'est
 * ce qui garantit qu'ils partiront vers l'API (`lotSaisiPour` en fait autant à
 * l'envoi), et donc qu'un retour à « En mon nom » soit enregistré.
 *
 * ⚠️ `saisi_pour_user_id` vaut `null` et non `''` — c'est un identifiant, et un
 * `''` envoyé à un champ entier serait refusé par le serveur.
 */
export function champsSaisiPour(
	objet: PorteSaisiPour | Record<string, unknown> | null | undefined,
): {
	saisi_pour_user_id: number | null;
	saisi_pour_nom: string;
	saisi_pour_email: string;
} {
	const o = (objet ?? {}) as PorteSaisiPour;
	return {
		saisi_pour_user_id: o.saisi_pour_user_id ?? null,
		saisi_pour_nom: o.saisi_pour_nom ?? '',
		saisi_pour_email: o.saisi_pour_email ?? '',
	};
}

import { comparerParNom } from '$lib/noms';

/** Une personne proposable dans le sélecteur « pour un résident inscrit ». */
export interface ResidentProposable {
	id: number;
	prenom: string;
	nom: string;
	email: string;
}

/**
 * Les résidents proposables — actifs, classés par nom de famille.
 *
 * ⚠️ **Ne lève jamais.** La section reste utilisable en « En mon nom » et pour
 * une personne extérieure : bloquer la saisie entière pour un défaut qui ne
 * concerne qu'un des trois modes coûterait plus qu'il ne protège.
 *
 * 🔴 Écrite ici parce que les trois formulaires en avaient besoin le même jour.
 * Recopiée, elle aurait divergé sur le tri — c'est précisément ce qui venait
 * d'arriver au classement des porteurs de badges.
 */
export async function chargerResidents(
	lister: () => Promise<unknown[]>,
): Promise<ResidentProposable[]> {
	try {
		const tous = (await lister()) as (ResidentProposable & { actif?: boolean })[];
		//  ⚠️ `comparerParNom` n'est PAS un paramètre : classer des personnes par
		//  nom de famille est la règle du site (`$lib/noms`), pas une option de
		//  l'appelant. La laisser choisir rouvrirait les cinq tris divergents que
		//  la centralisation vient de fermer.
		return tous.filter((u) => u.actif).sort(comparerParNom);
	} catch {
		return [];
	}
}

/** Ce qu'un objet déjà enregistré sait de son propriétaire. */
export interface PorteProprietaire {
	proprietaire_nom?: string | null;
	auteur_nom?: string | null;
}

/**
 * **À QUI part la copie** — le nom écrit dans « Envoyer une copie à … ».
 *
 * ## 🔴 Pourquoi cette fonction (15/09/2026)
 *
 * Signalé à l'écran :
 *
 * > « Quand on change "Saisi pour" avec une personne ayant un email, ça change
 * >   envoyer une copie à X »
 *
 * C'est le comportement **voulu** — arbitré le 12/09 : *« pour un ticket dont
 * le "Saisi pour" possède un résident inscrit ou une personne extérieure, ce
 * dernier se substitue à l'auteur »*. La copie va au PROPRIÉTAIRE, pas à celui
 * qui tape.
 *
 * ⚠️ Mais les écrans ne le disaient pas tous. `CarteTicket` et
 * `FilMessagesTicket` composaient `proprietaire_nom ?? auteur_nom` — deux fois
 * la même expression —, tandis que **les formulaires passaient `auteur_nom`
 * seul**. La même case annonçait donc deux noms différents selon l'écran, et
 * celui du formulaire était le mauvais.
 *
 * 🔴 Et en CRÉATION, aucun des deux ne convenait : rien n'est encore
 * enregistré, mais le rédacteur vient de désigner quelqu'un. La case doit dire
 * **ce qui partira**, pas ce qui était là — sinon elle promet un destinataire
 * et en sert un autre. C'est exactement ce que l'arbitrage du 31/08 reprochait
 * à « Envoyer une copie à l'auteur » : un libellé qui ne dit pas à qui l'on
 * écrit.
 *
 * @param objet L'objet enregistré, s'il existe.
 * @param saisie Le « Saisi pour » en cours de saisie, s'il y en a un. Il
 *   l'emporte : c'est lui qui sera enregistré au prochain clic.
 * @returns Le nom, ou `''` — jamais `undefined`. `CanauxNotification` retombe
 *   alors sur « l'auteur », jamais sur rien.
 */
export function nomCopie(
	objet: PorteProprietaire | null | undefined,
	saisie?: {
		mode: ModeSaisiPour;
		userId: number | null;
		nom: string;
		residents: ResidentProposable[];
	},
): string {
	if (saisie?.mode === 'exterieur' && saisie.nom.trim()) return saisie.nom.trim();
	if (saisie?.mode === 'resident' && saisie.userId != null) {
		const p = saisie.residents.find((r) => r.id === saisie.userId);
		//  ⚠️ Sans le résident sous la main (liste pas encore chargée), on retombe
		//  sur l'enregistré plutôt que d'afficher un nom vide : une case muette
		//  redeviendrait « Envoyer une copie à l'auteur », c'est-à-dire l'ambiguïté
		//  que le libellé nommé existe pour lever.
		if (p) return [p.prenom, (p.nom ?? '').toUpperCase()].filter(Boolean).join(' ');
	}
	return objet?.proprietaire_nom ?? objet?.auteur_nom ?? '';
}

/**
 * **L'état de saisie**, comme un objet — et non quatre variables parallèles.
 *
 * 🔴 Les quatre se déclaraient, se liaient et se passaient séparément dans
 * chaque formulaire : quatre `let`, quatre `bind:`, quatre arguments. Trois
 * écrans × huit lignes, pour UNE notion. Et deux de ces trois fichiers étaient
 * au plafond de modularité, si bien que la notion ne pouvait plus s'étendre.
 *
 * ⚠️ Les quatre champs changent **ensemble** — c'est déjà vrai côté serveur, où
 * ils voyagent groupés. Un objet le dit ; quatre variables laissent croire
 * qu'on peut en toucher une seule.
 */
export interface SaisieSaisiPour {
	mode: ModeSaisiPour;
	userId: number | null;
	nom: string;
	email: string;
}

/** L'état de saisie d'un objet existant — ou vierge. */
export function saisieDepuis(objet: PorteSaisiPour | null | undefined): SaisieSaisiPour {
	const c = champsSaisiPour(objet);
	return {
		mode: modeDepuis(objet),
		userId: c.saisi_pour_user_id,
		nom: c.saisi_pour_nom,
		email: c.saisi_pour_email,
	};
}

/** Ce qui part vers l'API, depuis l'état de saisie. */
export function lotDepuisSaisie(s: SaisieSaisiPour): Required<PorteSaisiPour> {
	return lotSaisiPour(s.mode, s.userId, s.nom, s.email);
}
